xbout/boutdataarray.py

```diff
@@ -133,6 +135,243 @@ def fromFieldAligned(self):
         result["direction_y"] = "Standard"
         return result
 
+    def fromRegion(self, region, with_guards=None):
+        """
+        Get a logically-rectangular section of data from a certain region.
+        Includes guard cells from neighbouring regions.
+
+        Parameters
+        ----------
+        region : str
+            Region to get data for
+        with_guards : int or dict of int, optional
+            Number of guard cells to include, by default use MXG and MYG from BOUT++.
+            Pass a dict to set different numbers for different coordinates.
+        """
+
+        region = self.data.regions[region]
+        xcoord = self.data.metadata['bout_xdim']
+        ycoord = self.data.metadata['bout_ydim']
+        keep_yboundaries = self.data.metadata['keep_yboundaries']
+
+        if with_guards is None:
+            mxg = self.data.metadata['MXG']
+            myg = self.data.metadata['MYG']
+        else:
+            try:
+                try:
+                    mxg = with_guards[xcoord]
+                except KeyError:
+                    mxg = self.data.metadata['MXG']
+                try:
+                    myg = with_guards[ycoord]
+                except KeyError:
+                    myg = self.data.metadata['MYG']
+            except TypeError:
+                mxg = with_guards
+                myg = with_guards
+
+        xslice, yslice = region.getSlices()
+        da = self.data.isel(**{xcoord: xslice, ycoord: yslice})
+
+        if mxg > 0:
```

Comment:
I think the refactor @TomNicholas suggested, pulling this logic into separate functions in `regions.py` has helped. The comments could probably do with improvement, but I struggling for what to add. The basic function is to fetch `mxg` (argument to `from_region`) x-guard cells and `myg` (argument to `from_region`) y-guard cells and stick them on the edges of the data from the requested region. The x-guard cells get a bit complicated, because of the possibility of a limiter topology where a region without y-boundaries fetches guard cells from a region with y-boundaries or vice-versa. Partly this is because I've chosen to handle boundary cells and guard cells differently in the `Region` code - boundary cells are always included (if they are present in the `Dataset`), whereas guard cells are included optionally, and different numbers can be requested in the call to `from_region()`. I think this is simpler than the alternative of treating the boundary cells as if they were guard cells though: in that case you'd have to handle the case when guard cells were requested but boundary cells were not present, and you'd have to be careful with the number of guard cells requested if you didn't want to lose the data from the boundary cells when splitting into regions.

Reply:
Ok, thanks! Yes, keeping a distinction between guard and boundary cells is a good idea.

Are the `mxg/myg > 0` conditionals necessary? I mean, is it possible/sensible to have `region.connection_inner not None` if `mxg == 0`?

Reply:
The `mxg>0` logic should probably be inside `_concat_inner_guards`, etc., rather than outside. `mxg=0` is allowed though, regardless of whether the connection is `None` or has a value, it just means don't get any guard cells from any connection.